spec/models/concerns/globalizable.rb

```diff
@@ -11,6 +11,15 @@
   let(:fields) { record.translated_attribute_names }
   let(:required_fields) { fields.select { |field| record.send(:required_attribute?, field) } }
   let(:attribute) { required_fields.sample || fields.sample }
+  let(:invalid_translation_attributes) do
+    Hash.new.tap do |translation_attributes|
+      translation_attributes[:id] = record.translations.find_by(locale: :es).id
+      translation_attributes[attribute] = ""
+      if attribute == :main_link_url && [:budget, :budget_phase].include?(factory_name)
+        translation_attributes[:main_link_text] = "Enlace"
+      end
+    end
+  end
```

Comment:
Why are these changes necessary? :thinking:

Reply:
Because `globalizable` specs do not take into account conditioned validations. The `main_link_url` field validation only takes place when the field `main_link_text` is fulfilled. The globalizable spec method that fulfils the required fields does not fulfil the `main_link_text`as it is not mandatory and then some spec crashes when executed for the `Budget` or `Budget::Phase` model. This is the fastest way I found to solve it. Maybe there is a better approach 🤔 .

Reply:
Two possible solutions come to mind :thinking:.

One would be defining an extra factory, like it was done for other models (see commit 66334b5757).

```ruby
  factory :budget do
  # ...
    factory :budget_with_link do
      main_link_text { "Link" }
      main_link_url { "/" }
    end
  end
 
  # Then, in the budget spec:
  it_behaves_like "globalizable", :budget_with_link
```

The other would be changing the way we mark translations as required in order to handle the case where translatable fields are required but only under certain conditions. Here's the patch:

```diff
diff --git a/app/models/concerns/globalizable.rb b/app/models/concerns/globalizable.rb
index 98613b5a7..4c47249bd 100644
--- a/app/models/concerns/globalizable.rb
+++ b/app/models/concerns/globalizable.rb
@@ -32,14 +32,9 @@ module Globalizable
     private
 
       def required_attribute?(attribute)
-        presence_validators = [ActiveModel::Validations::PresenceValidator,
-          ActiveRecord::Validations::PresenceValidator]
-
-        attribute_validators(attribute).any? { |validator| presence_validators.include? validator }
-      end
-
-      def attribute_validators(attribute)
-        self.class.validators_on(attribute).map(&:class)
+        self.class.validators_on(attribute).any? do |validator|
+          validator.kind == :presence && !validator.options[:unless]
+        end
       end
 
       def check_translations_number
```

And here's the test case for this patch :sweat_smile:. I had to add the `validate :check_translations_number, if: :translations_required?` because I didn't manage to add a test using the `on: :update` condition. Due to this, I'm not even sure we're testing the right scenario here :thinking:.

```ruby
require "rails_helper"

describe Globalizable do
  it "doesn't require a translation when translatable fields require conditional presence" do
    dummy_translation = Class.new do
      include ActiveModel::Model
      attr_accessor :title, :locale
      validates :locale, presence: true
    end

    stub_const("DummyTranslation", dummy_translation)

    dummy_model = Class.new do
      include ActiveModel::Model
      attr_accessor :title, :translations

      def self.globalize_accessors; end
      def self.accepts_nested_attributes_for(...); end
      def self.after_validation(...); end
      def self.paranoid?; end

      def self.translation_class
        DummyTranslation
      end

      def translated_attribute_names
        [:title]
      end

      include Globalizable
      validate :check_translations_number, if: :translations_required?
      validates_translation :title, presence: true, unless: -> { true }

      def translations
        @translations ||= []
      end
    end

    stub_const("DummyModel", dummy_model)

    expect(DummyModel.new).to be_valid
  end
end
```

The advantage of the second solution is we could remove the factories defined in commit 66334b5757. But even so, I'm not sure which one I like better, so up to you!